iml-services/iml-devices/src/breadth_first_parent_iterator.rs

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2020 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+use iml_wire_types::db::{Device, DeviceId};
+use std::collections::{BTreeMap, BTreeSet};
+
+pub struct BreadthFirstParentIterator<'a, 'b> {
+    devices: &'a BTreeMap<DeviceId, Device>,
+    parents: BTreeSet<DeviceId>,
+    next_parents: BTreeSet<DeviceId>,
+    _marker: &'b std::marker::PhantomData<()>,
+}
+
+impl<'a, 'b> BreadthFirstParentIterator<'a, 'b> {
+    pub fn new(devices: &'a BTreeMap<DeviceId, Device>, device_id: &'b DeviceId) -> Self {
+        tracing::trace!("Getting {:?} from devices", device_id);
+        // TODO: This is dangerous
+        let device = &devices[device_id];
```

Comment:
`devices` is a BTreeMap, so you could use `.get()` returning an `Option`. If the Option is None, the iterator will be depleted from the beginning (`.next() will return None), which is probably what we need, no?

Reply:
No.

Not finding a device is a caller mistake. Remember we're using `iterator.all(...)` to test a predicate on all returned devices. If iterator returns none, `all` will say `true`, which is incorrect.

It's the same reason we can't return `None` in `next` when we don't find a parent. Not finding a parent means tree is not consistent, not that we should stop iteration.

Strictly speaking the interface should be `-> Option<Result<...>>` which will make it insurmountable to use.

In practice this is never the case since we always start with device that's already in `devices`, and tree as given by `device-scanner` is always consistent and contains all parents.